src/libraries/System.Private.CoreLib/src/System/Index.cs

```diff
@@ -138,6 +143,15 @@ public override string ToString()
             return ((uint)Value).ToString();
         }
 
+        private void ThrowValueArgumentOutOfRange_NeedNonNegNumException()
+        {
+#if SYSTEM_PRIVATE_CORELIB
+            ThrowHelper.ThrowValueArgumentOutOfRange_NeedNonNegNumException();
```

Comment:
I seem to remember the JIT having some issues with this pattern, where it doesn't see the method it's being inlined as always-throws because that's nested inside of a helper. @EgorBo, am I misremembering?  I can't repro it now, so maybe I'm not remembering correctly or maybe it was fixed or maybe it's more complicated than I'm remembering?

Reply:
I'd be fine changing this method to throw the exception with the right message string if we wanted to do that instead of calling into the other throw helper.

Reply:
I've inlined the exception throws into these helpers to avoid breaking the throw helper pattern. Let me know if this looks good!

Reply:
> I seem to remember the JIT having some issues with this pattern, where it doesn't see the method it's being inlined as always-throws because that's nested inside of a helper. @EgorBo, am I misremembering? I can't repro it now, so maybe I'm not remembering correctly or maybe it was fixed or maybe it's more complicated than I'm remembering?

Just noticed the ping - I think it should work fine, the only case that might not work if these methods were generics and T was _Canon which is not the case here